lib/internal/promisify.js

```diff
@@ -0,0 +1,90 @@
+'use strict';
+
+const SUFFIX = 'Promise';
+const SHORTCUT_NAME = 'promised';
+const install = process._enablePromises ? _install : _noop;
+
+module.exports = {
+  normal,
+  single,
+  multiple
+};
+
+function normal(object, name) {
+  return _promisify(object, name, (lastArg, err, val) => {
+    if (err) {
+      if (lastArg &&
+          typeof lastArg === 'object' &&
+          err.code &&
+          typeof lastArg[err.code] === 'function') {
+        return lastArg[err.code]()
+      }
+      throw err;
+    }
+    return val;
+  });
+}
+
+function single(object, name) {
+  return _promisify(object, name, (lastArg, val) => {
+    return val;
+  });
+}
+
+function multiple(object, name, mapped) {
+  return _promisify(object, name, (lastArg, err, arg0, arg1) => {
+    if (err) {
+      if (lastArg &&
+          typeof lastArg === 'object' &&
+          err.code &&
+          typeof lastArg[err.code] === 'function') {
+        return lastArg[err.code]()
+      }
+      throw err;
+    }
+    return {
+      [mapped[0]]: arg0,
+      [mapped[1]]: arg1
+    };
+  });
+}
+
+function _promisify(object, name, handler) {
+  const fn = install(object, name, inner);
+
+  function inner() {
+    const args = Array.from(arguments);
+    const lastArg = args[args.length - 1];
+    return new Promise((resolve, reject) => {
+      fn.call(this, ...args, resolver);
+
+      function resolver(arg0, arg1, arg2) {
+        try {
+          resolve(handler(lastArg, arg0, arg1, arg2));
```

Comment:
@jfhbrook yes, promises normalize that - if you pass `JSON.parse` to a promise or async method you'll get consistent behavior `somePromiseFn().then(JSON.parse)` does not result in a synchronous throw - but rather an asynchronous rejection. If you don't handle it you get an `unhandledRejection` but it is recoverable. This also has the benefit of catching any places you forgot to deal the error case. For example, with promises you don't have to explicitly type `if(err) return cb(err);` (and forget the `return` :P). With promises every error path is checked and propagated - like in completely synchronous code.